File: luau_decompiler/codegen.py
"""
Code Generator: Converts AST nodes back into readable Luau source code.
"""

from . import ast_nodes as ast


class CodeGen:
    def __init__(self):
        self.indent_level = 0
        self.indent_str = "\t"

    def generate(self, node) -> str:
        if isinstance(node, ast.FunctionExpr) and node.name is None:
            # Main chunk - just emit body
            lines = []
            for stmt in node.body:
                lines.append(self._emit_stmt(stmt))
            return "\n".join(lines)
        return self._emit_expr(node)

    def _indent(self) -> str:
        return self.indent_str * self.indent_level

    def _emit_stmt(self, node) -> str:
        if isinstance(node, ast.LocalAssign):
            return self._emit_local_assign(node)
        elif isinstance(node, ast.Assign):
            return self._emit_assign(node)
        elif isinstance(node, ast.ReturnStat):
            return self._emit_return(node)
        elif isinstance(node, ast.ExprStat):
            return f"{self._indent()}{self._emit_expr(node.expr)}"
        elif isinstance(node, ast.IfStat):
            return self._emit_if(node)
        elif isinstance(node, ast.WhileStat):
            return self._emit_while(node)
        elif isinstance(node, ast.NumericForStat):
            return self._emit_numeric_for(node)
        elif isinstance(node, ast.GenericForStat):
            return self._emit_generic_for(node)
        elif isinstance(node, ast.DoBlock):
            return self._emit_do_block(node)
        elif isinstance(node, ast.BreakStat):
            return f"{self._indent()}break"
        elif isinstance(node, ast.ContinueStat):
            return f"{self._indent()}continue"
        elif isinstance(node, ast.Comment):
            return f"{self._indent()}-- {node.text}"
        elif isinstance(node, ast.RepeatStat):
            return self._emit_repeat(node)
        return f"{self._indent()}-- ?? {type(node).__name__}"

    def _emit_expr(self, node) -> str:
        if node is None:
            return "nil"
        if isinstance(node, ast.NilLiteral):
            return "nil"
        elif isinstance(node, ast.BoolLiteral):
            return "true" if node.value else "false"
        elif isinstance(node, ast.NumberLiteral):
            v = node.value
            if v == int(v) and abs(v) < 1e15:
                return str(int(v))
            return str(v)
        elif isinstance(node, ast.StringLiteral):
            return self._escape_string(node.value)
        elif isinstance(node, ast.VarArgExpr):
            return "..."
        elif isinstance(node, ast.LocalVar):
            return node.name
        elif isinstance(node, ast.UpvalueRef):
            return node.name
        elif isinstance(node, ast.GlobalRef):
            return node.name
        elif isinstance(node, ast.IndexExpr):
            return self._emit_index(node)
        elif isinstance(node, ast.MethodCall):
            return self._emit_method_call(node)
        elif isinstance(node, ast.FunctionCall):
            return self._emit_function_call(node)
        elif isinstance(node, ast.BinaryOp):
            return self._emit_binary_op(node)
        elif isinstance(node, ast.UnaryOp):
            return self._emit_unary_op(node)
        elif isinstance(node, ast.ConcatExpr):
            parts = [self._emit_expr(p) for p in node.parts]
            return " .. ".join(parts)
        elif isinstance(node, ast.TableConstructor):
            return self._emit_table(node)
        elif isinstance(node, ast.FunctionExpr):
            return self._emit_function_expr(node)
        elif isinstance(node, ast.IfExpr):
            cond = self._emit_expr(node.condition)
            then_expr = self._emit_expr(node.then_expr)
            else_expr = self._emit_expr(node.else_expr)
            return f"if {cond} then {then_expr} else {else_expr}"
        return f"--[[unknown: {type(node).__name__}]]"

    def _escape_string(self, s: str) -> str:
        escaped = s.replace("\\", "\\\\").replace('"', '\\"')
        escaped = escaped.replace("\n", "\\n").replace("\r", "\\r").replace("\t", "\\t")
        escaped = escaped.replace("\0", "\\0")
        return f'"{escaped}"'

    def _emit_index(self, node: ast.IndexExpr) -> str:
        obj = self._emit_expr(node.obj)
        if node.is_dot and isinstance(node.key, ast.StringLiteral):
            key = node.key.value
            if key.isidentifier() and not key.startswith("__"):
                return f"{obj}.{key}"
            return f'{obj}["{key}"]'
        key = self._emit_expr(node.key)
        return f"{obj}[{key}]"

    def _emit_method_call(self, node: ast.MethodCall) -> str:
        obj = self._emit_expr(node.obj)
        args = ", ".join(self._emit_expr(a) for a in node.args)
        return f"{obj}:{node.method}({args})"

    def _emit_function_call(self, node: ast.FunctionCall) -> str:
        func = self._emit_expr(node.func)
        args = ", ".join(self._emit_expr(a) for a in node.args)
        return f"{func}({args})"

    def _emit_binary_op(self, node: ast.BinaryOp) -> str:
        left = self._emit_expr(node.left)
        right = self._emit_expr(node.right)
        return f"({left} {node.op} {right})"

    def _emit_unary_op(self, node: ast.UnaryOp) -> str:
        operand = self._emit_expr(node.operand)
        if node.op == "#":
            return f"#{operand}"
        elif node.op == "-":
            return f"-{operand}"
        return f"{node.op} {operand}"

    def _emit_table(self, node: ast.TableConstructor) -> str:
        if not node.fields:
            return "{}"
        parts = []
        for f in node.fields:
            if isinstance(f, ast.TableField):
                if f.key is None:
                    parts.append(self._emit_expr(f.value))
                elif f.is_string_key and isinstance(f.key, ast.StringLiteral):
                    parts.append(f"{f.key.value} = {self._emit_expr(f.value)}")
                else:
                    parts.append(f"[{self._emit_expr(f.key)}] = {self._emit_expr(f.value)}")
            else:
                parts.append(self._emit_expr(f))
        if len(parts) <= 3:
            return "{ " + ", ".join(parts) + " }"
        inner = ",\n".join(f"{self._indent()}{self.indent_str}{p}" for p in parts)
        return "{\n" + inner + "\n" + self._indent() + "}"

    def _emit_function_expr(self, node: ast.FunctionExpr) -> str:
        params = ", ".join(node.params)
        if node.is_vararg:
            params = (params + ", ...") if params else "..."

        if not node.body:
            return f"function({params}) end"

        self.indent_level += 1
        body_lines = []
        for stmt in node.body:
            body_lines.append(self._emit_stmt(stmt))
        self.indent_level -= 1

        body_str = "\n".join(body_lines)
        return f"function({params})\n{body_str}\n{self._indent()}end"

    def _emit_local_assign(self, node: ast.LocalAssign) -> str:
        if (
            len(node.names) == 1
            and len(node.values) == 1
            and isinstance(node.values[0], ast.FunctionExpr)
        ):
            return self._emit_local_function(node.names[0], node.values[0])
        names = ", ".join(node.names)
        if node.values:
            vals = ", ".join(self._emit_expr(v) for v in node.values)
            return f"{self._indent()}local {names} = {vals}"
        return f"{self._indent()}local {names}"

    def _emit_assign(self, node: ast.Assign) -> str:
        if (
            len(node.targets) == 1
            and len(node.values) == 1
            and isinstance(node.values[0], ast.FunctionExpr)
        ):
            function_assign = self._emit_named_function(node.targets[0], node.values[0])
            if function_assign is not None:
                return function_assign
        targets = ", ".join(self._emit_expr(t) for t in node.targets)
        vals = ", ".join(self._emit_expr(v) for v in node.values)
        return f"{self._indent()}{targets} = {vals}"

    def _emit_local_function(self, name: str, node: ast.FunctionExpr) -> str:
        params = ", ".join(node.params)
        if node.is_vararg:
            params = (params + ", ...") if params else "..."
        lines = [f"{self._indent()}local function {name}({params})"]
        self.indent_level += 1
        for stmt in node.body:
            lines.append(self._emit_stmt(stmt))
        self.indent_level -= 1
        lines.append(f"{self._indent()}end")
        return "\n".join(lines)

    def _emit_named_function(self, target, node: ast.FunctionExpr):
        if not isinstance(target, ast.IndexExpr):
            return None
        if not target.is_dot or not isinstance(target.key, ast.StringLiteral):
            return None
        params = ", ".join(node.params)
        if node.is_vararg:
            params = (params + ", ...") if params else "..."
        head = f"{self._indent()}function {self._emit_index(target)}({params})"
        lines = [head]
        self.indent_level += 1
        for stmt in node.body:
            lines.append(self._emit_stmt(stmt))
        self.indent_level -= 1
        lines.append(f"{self._indent()}end")
        return "\n".join(lines)

    def _emit_return(self, node: ast.ReturnStat) -> str:
        if not node.values:
            return f"{self._indent()}return"
        vals = ", ".join(self._emit_expr(v) for v in node.values)
        return f"{self._indent()}return {vals}"

    def _emit_if(self, node: ast.IfStat) -> str:
        cond = self._emit_expr(node.condition)
        lines = [f"{self._indent()}if {cond} then"]
        self.indent_level += 1
        for s in node.then_body:
            lines.append(self._emit_stmt(s))
        self.indent_level -= 1
        for clause in node.elseif_clauses:
            ec, eb = clause
            lines.append(f"{self._indent()}elseif {self._emit_expr(ec)} then")
            self.indent_level += 1
            for s in eb:
                lines.append(self._emit_stmt(s))
            self.indent_level -= 1
        if node.else_body:
            lines.append(f"{self._indent()}else")
            self.indent_level += 1
            for s in node.else_body:
                lines.append(self._emit_stmt(s))
            self.indent_level -= 1
        lines.append(f"{self._indent()}end")
        return "\n".join(lines)

    def _emit_while(self, node: ast.WhileStat) -> str:
        cond = self._emit_expr(node.condition)
        lines = [f"{self._indent()}while {cond} do"]
        self.indent_level += 1
        for s in node.body:
            lines.append(self._emit_stmt(s))
        self.indent_level -= 1
        lines.append(f"{self._indent()}end")
        return "\n".join(lines)

    def _emit_repeat(self, node: ast.RepeatStat) -> str:
        lines = [f"{self._indent()}repeat"]
        self.indent_level += 1
        for s in node.body:
            lines.append(self._emit_stmt(s))
        self.indent_level -= 1
        lines.append(f"{self._indent()}until {self._emit_expr(node.condition)}")
        return "\n".join(lines)

    def _emit_numeric_for(self, node: ast.NumericForStat) -> str:
        start = self._emit_expr(node.start)
        limit = self._emit_expr(node.limit)
        step = self._emit_expr(node.step) if node.step else None
        header = f"for {node.var_name} = {start}, {limit}"
        if step:
            header += f", {step}"
        lines = [f"{self._indent()}{header} do"]
        self.indent_level += 1
        for s in node.body:
            lines.append(self._emit_stmt(s))
        self.indent_level -= 1
        lines.append(f"{self._indent()}end")
        return "\n".join(lines)

    def _emit_generic_for(self, node: ast.GenericForStat) -> str:
        vars_str = ", ".join(node.var_names)
        iters = ", ".join(self._emit_expr(it) for it in node.iterators)
        lines = [f"{self._indent()}for {vars_str} in {iters} do"]
        self.indent_level += 1
        for s in node.body:
            lines.append(self._emit_stmt(s))
        self.indent_level -= 1
        lines.append(f"{self._indent()}end")
        return "\n".join(lines)

    def _emit_do_block(self, node: ast.DoBlock) -> str:
        lines = [f"{self._indent()}do"]
        self.indent_level += 1
        for s in node.body:
            lines.append(self._emit_stmt(s))
        self.indent_level -= 1
        lines.append(f"{self._indent()}end")
        return "\n".join(lines)

File: luau_decompiler_go/codegen/codegen.go
// Package codegen converts AST nodes back into readable Luau source code.
package codegen

import (
	"fmt"
	"math"
	"strings"

	"Geckocompiler/ast"
)

// CodeGen transforms AST nodes into Luau source text.
type CodeGen struct {
	indentLevel int
	indentStr   string
}

// NewCodeGen creates a CodeGen with tab indentation.
func NewCodeGen() *CodeGen {
	return &CodeGen{indentStr: "\t"}
}

// Generate emits the full source for a top-level FunctionExpr.
func (g *CodeGen) Generate(node ast.Expr) string {
	if fe, ok := node.(*ast.FunctionExpr); ok && fe.Name == "" {
		lines := make([]string, 0, len(fe.Body))
		for _, stmt := range fe.Body {
			lines = append(lines, g.emitStmt(stmt))
		}
		return strings.Join(lines, "\n")
	}
	return g.emitExpr(node)
}

func (g *CodeGen) indent() string {
	return strings.Repeat(g.indentStr, g.indentLevel)
}

// --- Statement Emission ---

func (g *CodeGen) emitStmt(node ast.Stmt) string {
	switch s := node.(type) {
	case ast.LocalAssign:
		return g.emitLocalAssign(s)
	case ast.Assign:
		return g.emitAssign(s)
	case ast.ReturnStat:
		return g.emitReturn(s)
	case ast.ExprStat:
		return fmt.Sprintf("%s%s", g.indent(), g.emitExpr(s.Expr))
	case ast.IfStat:
		return g.emitIf(s)
	case ast.WhileStat:
		return g.emitWhile(s)
	case ast.NumericForStat:
		return g.emitNumericFor(s)
	case ast.GenericForStat:
		return g.emitGenericFor(s)
	case ast.DoBlock:
		return g.emitDoBlock(s)
	case ast.BreakStat:
		return g.indent() + "break"
	case ast.ContinueStat:
		return g.indent() + "continue"
	case ast.Comment:
		return fmt.Sprintf("%s-- %s", g.indent(), s.Text)
	case ast.RepeatStat:
		return g.emitRepeat(s)
	}
	return fmt.Sprintf("%s-- ?? %T", g.indent(), node)
}

// --- Expression Emission ---

func (g *CodeGen) emitExpr(node ast.Expr) string {
	if node == nil {
		return "nil"
	}
	switch e := node.(type) {
	case ast.NilLiteral:
		return "nil"
	case ast.BoolLiteral:
		if e.Value {
			return "true"
		}
		return "false"
	case ast.NumberLiteral:
		return formatNumber(e.Value)
	case ast.StringLiteral:
		return escapeString(e.Value)
	case ast.VarArgExpr:
		return "..."
	case ast.LocalVar:
		return e.Name
	case ast.UpvalueRef:
		return e.Name
	case ast.GlobalRef:
		return e.Name
	case ast.IndexExpr:
		return g.emitIndex(e)
	case ast.MethodCall:
		return g.emitMethodCall(e)
	case ast.FunctionCall:
		return g.emitFunctionCall(e)
	case ast.BinaryOp:
		return g.emitBinaryOp(e)
	case ast.UnaryOp:
		return g.emitUnaryOp(e)
	case ast.ConcatExpr:
		parts := make([]string, 0, len(e.Parts))
		for _, p := range e.Parts {
			parts = append(parts, g.emitExpr(p))
		}
		return strings.Join(parts, " .. ")
	case *ast.TableConstructor:
		return g.emitTable(e)
	case ast.TableConstructor:
		return g.emitTable(&e)
	case *ast.FunctionExpr:
		return g.emitFunctionExpr(e)
	case ast.FunctionExpr:
		return g.emitFunctionExpr(&e)
	case ast.IfExpr:
		return fmt.Sprintf("if %s then %s else %s",
			g.emitExpr(e.Condition), g.emitExpr(e.ThenExpr), g.emitExpr(e.ElseExpr))
	}
	return fmt.Sprintf("--[[unknown: %T]]", node)
}

func formatNumber(v float64) string {
	if v == math.Trunc(v) && math.Abs(v) < 1e15 {
		return fmt.Sprintf("%d", int64(v))
	}
	return fmt.Sprintf("%g", v)
}

func escapeString(s string) string {
	s = strings.ReplaceAll(s, "\\", "\\\\")
	s = strings.ReplaceAll(s, "\"", "\\\"")
	s = strings.ReplaceAll(s, "\n", "\\n")
	s = strings.ReplaceAll(s, "\r", "\\r")
	s = strings.ReplaceAll(s, "\t", "\\t")
	s = strings.ReplaceAll(s, "\x00", "\\0")
	return fmt.Sprintf("\"%s\"", s)
}

func (g *CodeGen) emitIndex(e ast.IndexExpr) string {
	obj := g.emitExpr(e.Obj)
	if e.IsDot {
		if sl, ok := e.Key.(ast.StringLiteral); ok {
			if isValidIdentifier(sl.Value) {
				return fmt.Sprintf("%s.%s", obj, sl.Value)
			}
			return fmt.Sprintf("%s[\"%s\"]", obj, sl.Value)
		}
	}
	return fmt.Sprintf("%s[%s]", obj, g.emitExpr(e.Key))
}

func isValidIdentifier(s string) bool {
	if s == "" || strings.HasPrefix(s, "__") {
		return false
	}
	for i, ch := range s {
		if i == 0 {
			if ch != '_' && (ch < 'a' || ch > 'z') && (ch < 'A' || ch > 'Z') {
				return false
			}
		} else {
			if ch != '_' && (ch < 'a' || ch > 'z') && (ch < 'A' || ch > 'Z') && (ch < '0' || ch > '9') {
				return false
			}
		}
	}
	return true
}

func (g *CodeGen) emitMethodCall(e ast.MethodCall) string {
	obj := g.emitExpr(e.Obj)
	args := make([]string, 0, len(e.Args))
	for _, a := range e.Args {
		args = append(args, g.emitExpr(a))
	}
	return fmt.Sprintf("%s:%s(%s)", obj, e.Method, strings.Join(args, ", "))
}

func (g *CodeGen) emitFunctionCall(e ast.FunctionCall) string {
	fn := g.emitExpr(e.Func)
	args := make([]string, 0, len(e.Args))
	for _, a := range e.Args {
		args = append(args, g.emitExpr(a))
	}
	return fmt.Sprintf("%s(%s)", fn, strings.Join(args, ", "))
}

func (g *CodeGen) emitBinaryOp(e ast.BinaryOp) string {
	return fmt.Sprintf("(%s %s %s)", g.emitExpr(e.Left), e.Op, g.emitExpr(e.Right))
}

func (g *CodeGen) emitUnaryOp(e ast.UnaryOp) string {
	operand := g.emitExpr(e.Operand)
	switch e.Op {
	case "#":
		return "#" + operand
	case "-":
		return "-" + operand
	}
	return fmt.Sprintf("%s %s", e.Op, operand)
}

func (g *CodeGen) emitTable(tc *ast.TableConstructor) string {
	if len(tc.Fields) == 0 {
		return "{}"
	}
	parts := make([]string, 0, len(tc.Fields))
	for _, f := range tc.Fields {
		if f.Key == nil {
			parts = append(parts, g.emitExpr(f.Value))
		} else if f.IsStringKey {
			if sk, ok := f.Key.(ast.StringLiteral); ok {
				parts = append(parts, fmt.Sprintf("%s = %s", sk.Value, g.emitExpr(f.Value)))
			} else {
				parts = append(parts, fmt.Sprintf("[%s] = %s", g.emitExpr(f.Key), g.emitExpr(f.Value)))
			}
		} else {
			parts = append(parts, fmt.Sprintf("[%s] = %s", g.emitExpr(f.Key), g.emitExpr(f.Value)))
		}
	}
	if len(parts) <= 3 {
		return "{ " + strings.Join(parts, ", ") + " }"
	}
	inner := make([]string, 0, len(parts))
	for _, p := range parts {
		inner = append(inner, fmt.Sprintf("%s%s%s", g.indent(), g.indentStr, p))
	}
	return "{\n" + strings.Join(inner, ",\n") + "\n" + g.indent() + "}"
}

func (g *CodeGen) emitFunctionExpr(fe *ast.FunctionExpr) string {
	params := strings.Join(fe.Params, ", ")
	if fe.IsVararg {
		if params != "" {
			params += ", ..."
		} else {
			params = "..."
		}
	}
	if len(fe.Body) == 0 {
		return fmt.Sprintf("function(%s) end", params)
	}
	g.indentLevel++
	bodyLines := make([]string, 0, len(fe.Body))
	for _, stmt := range fe.Body {
		bodyLines = append(bodyLines, g.emitStmt(stmt))
	}
	g.indentLevel--
	return fmt.Sprintf("function(%s)\n%s\n%send", params, strings.Join(bodyLines, "\n"), g.indent())
}

func (g *CodeGen) emitLocalAssign(s ast.LocalAssign) string {
	// Special case: local function
	if len(s.Names) == 1 && len(s.Values) == 1 {
		if fe, ok := s.Values[0].(*ast.FunctionExpr); ok {
			return g.emitLocalFunction(s.Names[0], fe)
		}
		if fe, ok := s.Values[0].(ast.FunctionExpr); ok {
			return g.emitLocalFunction(s.Names[0], &fe)
		}
	}
	names := strings.Join(s.Names, ", ")
	if len(s.Values) > 0 {
		vals := make([]string, 0, len(s.Values))
		for _, v := range s.Values {
			vals = append(vals, g.emitExpr(v))
		}
		return fmt.Sprintf("%slocal %s = %s", g.indent(), names, strings.Join(vals, ", "))
	}
	return fmt.Sprintf("%slocal %s", g.indent(), names)
}

func (g *CodeGen) emitAssign(s ast.Assign) string {
	// Special case: named function assignment
	if len(s.Targets) == 1 && len(s.Values) == 1 {
		var fe *ast.FunctionExpr
		if f, ok := s.Values[0].(*ast.FunctionExpr); ok {
			fe = f
		} else if f, ok := s.Values[0].(ast.FunctionExpr); ok {
			fe = &f
		}
		if fe != nil {
			if result := g.emitNamedFunction(s.Targets[0], fe); result != "" {
				return result
			}
		}
	}
	targets := make([]string, 0, len(s.Targets))
	for _, t := range s.Targets {
		targets = append(targets, g.emitExpr(t))
	}
	vals := make([]string, 0, len(s.Values))
	for _, v := range s.Values {
		vals = append(vals, g.emitExpr(v))
	}
	return fmt.Sprintf("%s%s = %s", g.indent(), strings.Join(targets, ", "), strings.Join(vals, ", "))
}

func (g *CodeGen) emitLocalFunction(name string, fe *ast.FunctionExpr) string {
	params := strings.Join(fe.Params, ", ")
	if fe.IsVararg {
		if params != "" {
			params += ", ..."
		} else {
			params = "..."
		}
	}
	lines := []string{fmt.Sprintf("%slocal function %s(%s)", g.indent(), name, params)}
	g.indentLevel++
	for _, stmt := range fe.Body {
		lines = append(lines, g.emitStmt(stmt))
	}
	g.indentLevel--
	lines = append(lines, g.indent()+"end")
	return strings.Join(lines, "\n")
}

func (g *CodeGen) emitNamedFunction(target ast.Expr, fe *ast.FunctionExpr) string {
	ie, ok := target.(ast.IndexExpr)
	if !ok || !ie.IsDot {
		return ""
	}
	if _, ok := ie.Key.(ast.StringLiteral); !ok {
		return ""
	}
	params := strings.Join(fe.Params, ", ")
	if fe.IsVararg {
		if params != "" {
			params += ", ..."
		} else {
			params = "..."
		}
	}
	head := fmt.Sprintf("%sfunction %s(%s)", g.indent(), g.emitIndex(ie), params)
	lines := []string{head}
	g.indentLevel++
	for _, stmt := range fe.Body {
		lines = append(lines, g.emitStmt(stmt))
	}
	g.indentLevel--
	lines = append(lines, g.indent()+"end")
	return strings.Join(lines, "\n")
}

func (g *CodeGen) emitReturn(s ast.ReturnStat) string {
	if len(s.Values) == 0 {
		return g.indent() + "return"
	}
	vals := make([]string, 0, len(s.Values))
	for _, v := range s.Values {
		vals = append(vals, g.emitExpr(v))
	}
	return fmt.Sprintf("%sreturn %s", g.indent(), strings.Join(vals, ", "))
}

func (g *CodeGen) emitIf(s ast.IfStat) string {
	lines := []string{fmt.Sprintf("%sif %s then", g.indent(), g.emitExpr(s.Condition))}
	g.indentLevel++
	for _, stmt := range s.ThenBody {
		lines = append(lines, g.emitStmt(stmt))
	}
	g.indentLevel--
	for _, ec := range s.ElseifClauses {
		lines = append(lines, fmt.Sprintf("%selseif %s then", g.indent(), g.emitExpr(ec.Condition)))
		g.indentLevel++
		for _, stmt := range ec.Body {
			lines = append(lines, g.emitStmt(stmt))
		}
		g.indentLevel--
	}
	if len(s.ElseBody) > 0 {
		lines = append(lines, g.indent()+"else")
		g.indentLevel++
		for _, stmt := range s.ElseBody {
			lines = append(lines, g.emitStmt(stmt))
		}
		g.indentLevel--
	}
	lines = append(lines, g.indent()+"end")
	return strings.Join(lines, "\n")
}

func (g *CodeGen) emitWhile(s ast.WhileStat) string {
	lines := []string{fmt.Sprintf("%swhile %s do", g.indent(), g.emitExpr(s.Condition))}
	g.indentLevel++
	for _, stmt := range s.Body {
		lines = append(lines, g.emitStmt(stmt))
	}
	g.indentLevel--
	lines = append(lines, g.indent()+"end")
	return strings.Join(lines, "\n")
}

func (g *CodeGen) emitRepeat(s ast.RepeatStat) string {
	lines := []string{g.indent() + "repeat"}
	g.indentLevel++
	for _, stmt := range s.Body {
		lines = append(lines, g.emitStmt(stmt))
	}
	g.indentLevel--
	lines = append(lines, fmt.Sprintf("%suntil %s", g.indent(), g.emitExpr(s.Condition)))
	return strings.Join(lines, "\n")
}

func (g *CodeGen) emitNumericFor(s ast.NumericForStat) string {
	start := g.emitExpr(s.Start)
	limit := g.emitExpr(s.Limit)
	header := fmt.Sprintf("for %s = %s, %s", s.VarName, start, limit)
	if s.Step != nil {
		header += ", " + g.emitExpr(s.Step)
	}
	lines := []string{fmt.Sprintf("%s%s do", g.indent(), header)}
	g.indentLevel++
	for _, stmt := range s.Body {
		lines = append(lines, g.emitStmt(stmt))
	}
	g.indentLevel--
	lines = append(lines, g.indent()+"end")
	return strings.Join(lines, "\n")
}

func (g *CodeGen) emitGenericFor(s ast.GenericForStat) string {
	varsStr := strings.Join(s.VarNames, ", ")
	iters := make([]string, 0, len(s.Iterators))
	for _, it := range s.Iterators {
		iters = append(iters, g.emitExpr(it))
	}
	lines := []string{fmt.Sprintf("%sfor %s in %s do", g.indent(), varsStr, strings.Join(iters, ", "))}
	g.indentLevel++
	for _, stmt := range s.Body {
		lines = append(lines, g.emitStmt(stmt))
	}
	g.indentLevel--
	lines = append(lines, g.indent()+"end")
	return strings.Join(lines, "\n")
}

func (g *CodeGen) emitDoBlock(s ast.DoBlock) string {
	lines := []string{g.indent() + "do"}
	g.indentLevel++
	for _, stmt := range s.Body {
		lines = append(lines, g.emitStmt(stmt))
	}
	g.indentLevel--
	lines = append(lines, g.indent()+"end")
	return strings.Join(lines, "\n")
}
